File: projects/spring-boot-restv2/demo/src/main/java/com/sourabh/demo/repository/TransactionRepository.java
package com.sourabh.demo.repository;

import com.sourabh.demo.model.Transaction;
import org.bson.Document;
import org.springframework.data.mongodb.core.aggregation.AggregationResults;
import org.springframework.data.mongodb.repository.Aggregation;
import org.springframework.data.mongodb.repository.MongoRepository;

public interface TransactionRepository extends MongoRepository<Transaction,Integer> {
    @Aggregation(pipeline = {"{$group:{_id:'$customer.type',sum:{$sum:'$amount' }}}","{$sort:{sum:-1}}"})
    AggregationResults<Document> getRoomWiseEarnings();
}

File: projects/spring-boot-rest-kt/hotel-reservation/src/main/kotlin/com/sourabh/hotelreservation/repository/TransactionRepository.kt
package com.sourabh.hotelreservation.repository

import com.sourabh.hotelreservation.model.Transaction
import org.bson.Document
import org.springframework.data.mongodb.core.aggregation.AggregationResults
import org.springframework.data.mongodb.repository.Aggregation
import org.springframework.data.mongodb.repository.MongoRepository
import org.springframework.stereotype.Repository

@Repository
interface TransactionRepository :MongoRepository<Transaction, Int> {
    @Aggregation(pipeline = ["{\$group:{_id:'\$customer.type',sum:{\$sum:'\$amount' }}}", "{\$sort:{sum:-1}}"])
    fun getRoomWiseEarnings(): AggregationResults<Document?>?
}